opam var package var not found

`opam config var` find the `installed` variable value, but not `opam var`
```
% opam var ocp-index:installed
[ERROR] Variable ocp-index:installed not found

% opam config var ocp-index:installed
true
```

--- a/master_changes.md
+++ b/master_changes.md
@@ -36,7 +36,6 @@ Possibly scripts breaking changes are prefixed with ✘
   * Refactor package status [#4152 @rjbou]
   * Add Macport support [#4152 @rjbou]
 
-
 ## Remove
   * Fix autoremove env var handling [#4219 @rjbou - fix #4217]
 
@@ -52,3 +51,8 @@ Possibly scripts breaking changes are prefixed with ✘
 
 ## Doc
   * add doc/warning for  Filename.rmdir_cleanup [#4197 @rjbou]
+
+## Var
+  * Not found message show scope [#4192 @rjbou]
+  * No scope needed for variable display [#4192 @rjbou - fix #4183]
+  * Fix package variable resolution [#4192 @rjbou - fix #4182]

--- a/src/client/opamCommands.ml
+++ b/src/client/opamCommands.ml
@@ -847,7 +847,7 @@ module Var_Option_Common = struct
       else match var with
         | None -> `All
         | Some f -> match cmd with
-          | `var -> `Switch
+          | `var -> `All_var
           | `option -> OpamConfigCommand.get_scope f
     in
     let apply =
@@ -869,6 +869,16 @@ module Var_Option_Common = struct
        | `var -> OpamConfigCommand.vars_list gt
        | `option -> OpamConfigCommand.options_list gt);
       `Ok ()
+    | `All_var ->
+      (match apply with
+       | `value_wo_eq v ->
+         OpamGlobalState.with_ `Lock_none @@ fun gt ->
+         OpamConfigCommand.var_show gt v;
+         `Ok ()
+       | `empty -> assert false (* can't happen *)
+       | `value_eq _ ->
+         `Error (true, "variable setting needs a scope, \
+                        use '--global' or '--switch <switch>'"))
     | (`Global | `Switch) as scope ->
       match cmd, apply with
       | _ , `empty ->
@@ -1140,7 +1150,7 @@ let config =
       `Ok (OpamConfigCommand.expand gt str)
     | Some `var, [var] ->
       OpamGlobalState.with_ `Lock_none @@ fun gt ->
-      (try `Ok (OpamConfigCommand.variable gt var)
+      (try `Ok (OpamConfigCommand.var_show gt var)
        with Failure msg -> `Error (false, msg))
     | Some `subst, (_::_ as files) ->
       OpamGlobalState.with_ `Lock_none @@ fun gt ->

--- a/src/client/opamConfigCommand.ml
+++ b/src/client/opamConfigCommand.ml
@@ -945,6 +945,7 @@ let is_switch_defined_var switch_config v =
     (OpamVariable.of_string v) <> None
   || (try let _path = OpamTypesBase.std_path_of_string v in true
       with Failure _ -> false)
+  || OpamStd.String.contains_char v ':'
 
 let var_switch_raw ?(only_switch=true) gt v =
   match OpamStateConfig.get_switch_opt () with
@@ -971,10 +972,6 @@ let var_show_switch gt ?st v =
     let resolve_switch st =
       if is_switch_defined_var st.switch_config v then
         var_show (OpamPackageVar.resolve st) v
-      else
-        (* maybe package one *)
-      if OpamStd.String.contains_char v ':' then
-        var_show (OpamPackageVar.resolve_switch st) v
       else
         OpamConsole.error_and_exit `Not_found "Variable %s not found" v
     in

--- a/src/client/opamConfigCommand.ml
+++ b/src/client/opamConfigCommand.ml
@@ -933,7 +933,7 @@ let option_show_switch gt ?st field =
 let option_show_global gt field =
   option_show OpamFile.Config.to_list (confset_global gt) field
 
-let var_show resolve v =
+let var_show_t resolve v =
   match resolve (OpamVariable.Full.of_string v) with
   | Some c ->
     OpamConsole.msg "%s\n" (OpamVariable.string_of_variable_contents c)
@@ -971,7 +971,7 @@ let var_show_switch gt ?st v =
   if var_switch_raw gt v = None then
     let resolve_switch st =
       if is_switch_defined_var st.switch_config v then
-        var_show (OpamPackageVar.resolve st) v
+        var_show_t (OpamPackageVar.resolve st) v
       else
         OpamConsole.error_and_exit `Not_found "Variable %s not found" v
     in
@@ -979,13 +979,12 @@ let var_show_switch gt ?st v =
     | Some st -> resolve_switch st
     | None -> OpamSwitchState.with_ `Lock_none gt resolve_switch
 
-let var_show_global gt = var_show (OpamPackageVar.resolve_global gt)
+let var_show_global gt = var_show_t (OpamPackageVar.resolve_global gt)
 
-(* deprecated, kept for `opam config var` retro-compatibility *)
-let variable gt v =
+let var_show gt v =
   if var_switch_raw ~only_switch:false gt v = None then
     OpamSwitchState.with_ `Lock_none gt @@ fun st ->
-    var_show (OpamPackageVar.resolve st) v
+    var_show_t (OpamPackageVar.resolve st) v
 
 (* detect scope *)
 let get_scope field =

--- a/src/client/opamConfigCommand.ml
+++ b/src/client/opamConfigCommand.ml
@@ -933,12 +933,15 @@ let option_show_switch gt ?st field =
 let option_show_global gt field =
   option_show OpamFile.Config.to_list (confset_global gt) field
 
-let var_show_t resolve v =
+let var_show_t resolve ?switch v =
   match resolve (OpamVariable.Full.of_string v) with
   | Some c ->
     OpamConsole.msg "%s\n" (OpamVariable.string_of_variable_contents c)
   | None ->
-    OpamConsole.error_and_exit `Not_found "Variable %s not found" v
+    OpamConsole.error_and_exit `Not_found "Variable %s not found in %s" v
+      (match switch with
+           | None -> "global config"
+           | Some switch -> "in switch " ^ (OpamSwitch.to_string switch))
 
 let is_switch_defined_var switch_config v =
   OpamFile.Switch_config.variable switch_config
@@ -971,20 +974,25 @@ let var_show_switch gt ?st v =
   if var_switch_raw gt v = None then
     let resolve_switch st =
       if is_switch_defined_var st.switch_config v then
-        var_show_t (OpamPackageVar.resolve st) v
+        var_show_t (OpamPackageVar.resolve st) ~switch:st.switch v
       else
-        OpamConsole.error_and_exit `Not_found "Variable %s not found" v
+        OpamConsole.error_and_exit `Not_found
+          "Variable %s not found in switch %s"
+          v (OpamSwitch.to_string st.switch)
     in
     match st with
     | Some st -> resolve_switch st
     | None -> OpamSwitchState.with_ `Lock_none gt resolve_switch
 
-let var_show_global gt = var_show_t (OpamPackageVar.resolve_global gt)
+let var_show_global gt f = var_show_t (OpamPackageVar.resolve_global gt) f
 
 let var_show gt v =
   if var_switch_raw ~only_switch:false gt v = None then
     OpamSwitchState.with_ `Lock_none gt @@ fun st ->
-    var_show_t (OpamPackageVar.resolve st) v
+    let switch =
+      if is_switch_defined_var st.switch_config v then Some st.switch else None
+    in
+    var_show_t (OpamPackageVar.resolve st) ?switch v
 
 (* detect scope *)
 let get_scope field =

--- a/src/client/opamConfigCommand.mli
+++ b/src/client/opamConfigCommand.mli
@@ -37,10 +37,6 @@ val print_eval_env: csh:bool -> sexp:bool -> fish:bool -> env -> unit
 (** Display the content of all available packages variables *)
 val list: 'a switch_state -> name list -> unit
 
-(** Display the content of a given variable. Load a switch state only if the
-    variable is not found on raw state. *)
-val variable: 'a global_state -> string -> unit
-
 (** Substitute files *)
 val subst: 'a global_state -> basename list -> unit
 
@@ -147,3 +143,4 @@ val option_show_switch:
 val var_show_global: 'a global_state -> string -> unit
 val var_show_switch:
   'a global_state -> ?st:'b switch_state -> string -> unit
+val var_show: 'a global_state -> string -> unit

--- a/src/client/opamConfigCommand.mli
+++ b/src/client/opamConfigCommand.mli
@@ -53,7 +53,7 @@ val exec:
 (** Functions handling `opam var` and `opam option` command *)
 
 (** Given an `opam option` field or field-value argument, detect the scope,
-    switch, global or inexistent field
+    switch, global or nonexistent field
     (cf. [OpamCommands.Var_Option_Common.var_option]) *)
 val get_scope: string -> [> `Switch | `Global | `None of string ]
 

